airgen/src/lib.rs

```diff
@@ -20,71 +21,205 @@ use itertools::Itertools;
+    abs_to_loc: BTreeMap<AbsoluteSymbolPath, Location>,
+    map: BTreeMap<Location, Instance>,
+}
+
+impl Instances {
+    /// Trim to remove instances unreachable from a given Location
+    /// Returns the instance map
+    /// untested!
+    fn trim(self, from: &Location) -> BTreeMap<Location, Instance> {
+        let mut reached: BTreeSet<Location> = Default::default();
+        let mut queue = vec![from];
+        let mut map = self.map;
+        while let Some(location) = queue.pop() {
+            // `location` is reached
+            reached.insert(location.clone());
+            let instance = &map.get(location).unwrap();
+            // visit all params
+            queue.extend(&instance.members);
+        }
+        // remove all keys that aren't reached
+        map.retain(|l, _| reached.contains(l));
+        map
+    }
+
+    fn fold_instance(
+        &mut self,
+        file: &AnalysisASMFile,
+        location: Location,
+        instance: &MachineInstance,
+    ) -> Location {
+        match &instance.value {
+            MachineInstanceExpression::Value(v) => {
+                let ty = file.machine(&instance.ty);
+                assert_eq!(ty.params.0.len() + ty.submachines.len(), v.len());
+
+                let members = ty
+                    .params
+                    .0
+                    .iter()
+                    .map(|param| &param.name)
+                    .chain(ty.submachines.iter().map(|d| &d.name))
+                    .zip(v)
+                    .map(|(name, instance)| {
+                        self.fold_instance(file, location.clone().join(name.clone()), instance)
+                    })
+                    .collect();
+                self.map.insert(
+                    location.clone(),
+                    Instance {
+                        ty: instance.ty.clone(),
+                        members,
+                    },
+                );
+                location.clone()
+            }
+            MachineInstanceExpression::Reference(r) => self.abs_to_loc[r].clone(),
+        }
+    }
+}
+
+/// Instantiate machine type at `ty_path` by instantiating all submachines recursively
+fn instantiate(
+    input: &AnalysisASMFile,
+    instances: &mut BTreeMap<AbsoluteSymbolPath, MachineInstance>,
+    path: &AbsoluteSymbolPath,
+    ty_path: &AbsoluteSymbolPath,
+    args: &Vec<AbsoluteSymbolPath>,
+) {
+    let ty = input.machine(ty_path);
+
+    assert_eq!(ty.params.0.len(), args.len());
 
+    let submachines: Vec<MachineInstance> = ty
+        .params
+        .0
+        .iter()
+        .zip(args)
+        .map(|(param, argument)| {
+            let ty = AbsoluteSymbolPath::default().join(param.ty.clone().unwrap());
+            match input.items.get(&ty) {
+                Some(Item::Machine(_)) => MachineInstance {
+                    ty,
+                    value: MachineInstanceExpression::Reference(argument.clone()),
+                },
+                _ => unimplemented!(),
```

Comment:
is this unreachable instead because of a previous check (i don't recall)? this is a reference to a non-existent machine type right?

Reply:
Do we expect any other types of parameters here? My understanding is that we only ever get machines as arguments.
Now assuming we get a machine, I'm not sure we currently enforce that this machine must be defined somewhere. This is something we need to do in `machine_check`.

I'm removing this check here entirely, as we make this assumption in a many places in this code, for example in `fold_instance`.